qutip/core/data/add.pyx

```diff
@@ -51,6 +52,10 @@ cdef idxint _add_csr(Accumulator *acc, CSR a, CSR b, CSR c) nogil:
     """
     cdef idxint row, ptr_a, ptr_b, ptr_a_max, ptr_b_max, nnz=0, col_a, col_b
     cdef idxint ncols = a.shape[1]
+    cdef double tol
+    with gil:
+        if settings.core['auto_tidyup']:
+            tol = settings.core['auto_tidyup_atol']
```

Comment:
It would be better to pass in the tolerance value in the call from `add_csr`, which holds the GIL, rather than playing yoyo with it here (and below).